### PR TITLE
refactor: Refactor data scratch-softwarecatalog

### LINKED ISSUE #1751: Data scratch - software catalog
# Plan

According to the specification described in epic [#1727](https://github.com/Flank/flank/issues/1727)
add abstraction and implementation for the following part of the data layer:

## Abstraction

`ftl/data/SoftwareCatalog.kt`

```kotlin
package ftl.data

val fetchSoftwareCatalog: SoftwareCatalog.Fetch get() = TODO()

data class SoftwareCatalog(
    val orchestratorVersion: String
) {
    interface Fetch : () -> SoftwareCatalog
}
```

## Target

* `ftl/environment/ProvidedSoftwareCatalog.kt`

## Adapter

`ftl/adapter/GoogleSoftwareCatalogFetch.kt`

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #1751 

Code is refactored according to the description in #1751 
Tests work as intended.

## Checklist
- [x] Unit tested

### CODE REVIEW

**Review by pawelpasterz — APPROVED** [review]

File: test_runner/src/main/kotlin/ftl/adapter/GoogleProvidedSoftwareCatalog.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    SoftwareCatalog.Fetch,
+    () -> ProvidedSoftwareCatalog by {
+        getProvidedSoftwareToApiModel()
+    }
+
+object GoogleProvidedSoftwareCatalogAsTable :
+    SoftwareCatalog.FetchAsTable,
+    () -> String by {
+        getProvidedSoftwareAsATableToApiModel()
+    }
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/adapter/GoogleProvidedSoftwareCatalog.kt:18]:
```suggestion
```
Parsing to the table is not the responsibility of this layer

File: test_runner/src/main/kotlin/ftl/adapter/google/SoftwareCatalogAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package ftl.adapter.google
+
+import ftl.client.google.getProvidedSoftware
+import ftl.client.google.providedSoftwareAsTable
+
+internal fun getProvidedSoftwareToApiModel() = getProvidedSoftware()
+internal fun getProvidedSoftwareAsATableToApiModel() = providedSoftwareAsTable()
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/adapter/google/SoftwareCatalogAdapter.kt:7]:
```suggestion
```

File: test_runner/src/main/kotlin/ftl/api/SoftwareCatalog.kt [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import ftl.adapter.GoogleProvidedSoftwareCatalogAsTable
+
+val fetchSoftwareCatalog: SoftwareCatalog.Fetch get() = GoogleProvidedSoftwareCatalog
+val fetchSoftwareCatalogAsTable: SoftwareCatalog.FetchAsTable get() = GoogleProvidedSoftwareCatalogAsTable
+
+data class SoftwareCatalog(
+    val orchestratorVersion: String
+) {
+    interface Fetch : () -> ProvidedSoftwareCatalog
+    interface FetchAsTable : () -> String
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/api/SoftwareCatalog.kt:14]:
```suggestion
```

File: test_runner/src/main/kotlin/ftl/api/SoftwareCatalog.kt [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package ftl.api
+
+import com.google.testing.model.ProvidedSoftwareCatalog
+import ftl.adapter.GoogleProvidedSoftwareCatalog
+import ftl.adapter.GoogleProvidedSoftwareCatalogAsTable
+
+val fetchSoftwareCatalog: SoftwareCatalog.Fetch get() = GoogleProvidedSoftwareCatalog
+val fetchSoftwareCatalogAsTable: SoftwareCatalog.FetchAsTable get() = GoogleProvidedSoftwareCatalogAsTable
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/api/SoftwareCatalog.kt:8]:
```suggestion
```

File: test_runner/src/main/kotlin/ftl/domain/DescribeAndroidTestEnvironment.kt [hunk 5]
```diff
@@ -4,11 +4,11 @@ import flank.common.logLn
 import ftl.api.Platform
 import ftl.api.fetchIpBlocks
 import ftl.api.fetchOrientation
+import ftl.api.fetchSoftwareCatalogAsTable
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/domain/DescribeAndroidTestEnvironment.kt:7]:
```suggestion
```

File: test_runner/src/main/kotlin/ftl/domain/DescribeIosTestEnvironment.kt [hunk 6]
```diff
@@ -4,10 +4,10 @@ import flank.common.logLn
 import ftl.api.Platform
 import ftl.api.fetchIpBlocks
 import ftl.api.fetchOrientation
+import ftl.api.fetchSoftwareCatalogAsTable
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/domain/DescribeIosTestEnvironment.kt:7]:
```suggestion
```

**Review by jan-gogo — DISMISSED** [review]

**Comment by jan-gogo** [comment]
@Sloox 
Check this example:
https://github.com/Flank/flank/pull/1819/files#diff-9dab19db89ea08ff978778342185b8042a604f936ed9e649407972850aa6ccf0R12
this is properly solved logging at this stage of refactoring.

**Comment by Sloox** [comment]
@jan-gogo Changes applied.

File: test_runner/src/main/kotlin/ftl/domain/DescribeAndroidTestEnvironment.kt [hunk 7]
```diff
@@ -21,7 +22,7 @@ fun DescribeAndroidTestEnvironment.invoke() {
     logLn(AndroidCatalog.devicesCatalogAsTable(projectId))
     logLn(AndroidCatalog.supportedVersionsAsTable(projectId))
     logLn(AndroidCatalog.localesAsTable(projectId))
-    logLn(providedSoftwareAsTable())
+    logLn(fetchSoftwareCatalog().asTable())
```
Comment (adamfilipow92) [inline test_runner/src/main/kotlin/ftl/domain/DescribeAndroidTestEnvironment.kt:25]: Nothing important but other methods converting data to the table was named ```.toCliTable()``` maybe we can keep one naming conversion? WDYT?

Comment (Sloox) [reply test_runner/src/main/kotlin/ftl/domain/DescribeAndroidTestEnvironment.kt:25]: @adamfilipow92 sure.

**Review by adamfilipow92 — APPROVED** [review]
👍

**Pull request merged**